CS4001/4042 Assignment 1, Part B, Q4
---

Model degradation is a common issue faced when deploying machine learning models (including neural networks) in the real world. New data points could exhibit a different pattern from older data points due to factors such as changes in government policy or market sentiments. For instance, housing prices in Singapore have been increasing and the Singapore government has introduced 3 rounds of cooling measures over the past years (16 December 2021, 30 September 2022, 27 April 2023).

In such situations, the distribution of the new data points could differ from the original data distribution which the models were trained on. Recall that machine learning models often work with the assumption that the test distribution should be similar to train distribution. When this assumption is violated, model performance will be adversely impacted.  In the last part of this assignment, we will investigate to what extent model degradation has occurred.




---



---



Your co-investigators used a linear regression model to rapidly test out several combinations of train/test splits and shared with you their findings in a brief report attached in Appendix A below. You wish to investigate whether your deep learning model corroborates with their findings.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
!pip install alibi-detect
!pip install pytorch_tabular[extra]

In [3]:
SEED = 42

import os

import random
random.seed(SEED)

import numpy as np
np.random.seed(SEED)

import pandas as pd

from alibi_detect.cd import TabularDrift

> Evaluate your model from B1 on data from year 2022 and report the test R2.

In [4]:
from pytorch_tabular import TabularModel
import torch
df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/hdb_price_prediction.csv')

# TODO: Enter your code here
model = TabularModel.load_from_checkpoint("/content/drive/MyDrive/Colab Notebooks/NNDL_partb")

test_data1 = df[df['year'] == 2022]

result = model.evaluate(test_data1)[0]
pred_df = model.predict(test_data1)
print(result)

test_loss = result['test_loss']
test_mse = result['test_mean_squared_error']
test_target_values = test_data1["resale_price"]
pred_values = pred_df["resale_price_prediction"]
test_r2 = 1 - (test_mse / np.var(test_target_values))
print(f"Test R^2 on data from 2022: {test_r2}")

/usr/local/lib/python3.10/dist-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: '/usr/local/lib/python3.10/dist-packages/torchvision/image.so: undefined symbol: _ZN3c104cuda20CUDACachingAllocator9allocatorE'If you don't plan on using image functionality from `torchvision.io`, you can ignore this warning. Otherwise, there might be something wrong with your environment. Did you have `libjpeg` or `libpng` installed before building `torchvision` from source?
  warn(
2023-10-14 15:14:40,318 - {pytorch_tabular.tabular_model:129} - INFO - Experiment Tracking is turned off
INFO:pytorch_tabular.tabular_model:Experiment Tracking is turned off
2023-10-14 15:14:40,330 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
INFO:pytorch_tabular.tabular_model:Preparing the Trainer
/usr/local/lib/python3.10/dist-packages/pytorch_lightning/trainer/connectors/accelerator_connector.py:589: LightningDeprecationWarning: The Trainer argument `auto_select

Output()

/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called 
`self.log('test_loss', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called 
`self.log('test_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_loss         │       16267038720.0       │
│  test_mean_squared_error  │       16267038720.0       │
└───────────────────────────┴───────────────────────────┘

Output()

{'test_loss': 16267038720.0, 'test_mean_squared_error': 16267038720.0}
Test R^2 on data from 2022: 0.43884723246089874


> Evaluate your model from B1 on data from year 2023 and report the test R2.

In [5]:
# TODO: Enter your code here
test_data2 = df[df['year'] == 2023]
result = model.evaluate(test_data2)[0]
pred_df = model.predict(test_data2)
print(result)

test_loss = result['test_loss']
test_mse = result['test_mean_squared_error']
test_target_values = test_data2["resale_price"]
pred_values = pred_df["resale_price_prediction"]
test_r2 = 1 - (test_mse / np.var(test_target_values))
print(f"Test R^2 on data from 2023: {test_r2}")

Output()

/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called 
`self.log('test_loss', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called 
`self.log('test_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_loss         │       24701362176.0       │
│  test_mean_squared_error  │       24701362176.0       │
└───────────────────────────┴───────────────────────────┘

Output()

{'test_loss': 24701362176.0, 'test_mean_squared_error': 24701362176.0}
Test R^2 on data from 2023: 0.16212599113143344


> Did model degradation occur for the deep learning model?


\# TODO: \  
Yes, model degradation has occurred.  
Test R^2 on data from 2022: 0.43884723246089874  
Test R^2 on data from 2023: 0.16212599113143344  
Since, R^2 has decreased from 2022 to 2023, model degradation has occurred.



---



---



Model degradation could be caused by [various data distribution shifts](https://huyenchip.com/2022/02/07/data-distribution-shifts-and-monitoring.html#data-shift-types): covariate shift (features), label shift and/or concept drift (altered relationship between features and labels).
There are various conflicting terminologies in the [literature](https://www.sciencedirect.com/science/article/pii/S0950705122002854#tbl1). Let’s stick to this reference for this assignment.

> Using the **Alibi Detect** library, apply the **TabularDrift** function with the training data (year 2019 and before) used as the reference and **detect which features have drifted** in the 2023 test dataset. Before running the statistical tests, ensure you **sample 1000 data points** each from the train and test data. Do not use the whole train/test data. (Hint: use this example as a guide https://docs.seldon.io/projects/alibi-detect/en/stable/examples/cd_chi2ks_adult.html)


In [6]:
train_data = df[df['year'] <= 2019].sample(1000)
test_data = df[df['year'] == 2023].sample(1000)

train_data_numpy = train_data.values  # Convert 'train_data' to a NumPy array
test_data_numpy = test_data.values  # Convert 'test_data' to a NumPy array
# # Define the indices for categorical columns
col_idx = {
    'month': 0,
    'town': 2,
    'flat_model_type': 9,
    'storey_range': 12}

# Define the indices for categorical columns
categories_per_feature = {
    0: None,
    2: None,
    9: None,
    12: None}

# Create the TabularDrift detector
drift_detector = TabularDrift(train_data_numpy, p_val=0.05, categories_per_feature=categories_per_feature)
preds = drift_detector.predict(test_data_numpy)
labels = ['No!', 'Yes!']
print(preds)

{'data': {'is_drift': 1, 'distance': array([4.3622476e+02, 1.0000000e+00, 3.9025555e+01, 4.1000001e-02,
       5.7000000e-02, 6.8999998e-02, 6.3000001e-02, 2.3000000e-02,
       3.9000001e-02, 5.9055187e+01, 1.4700000e-01, 8.6999997e-02,
       2.1271683e+01, 3.9500001e-01], dtype=float32), 'p_val': array([0.0000000e+00, 0.0000000e+00, 3.6623273e-02, 3.6023447e-01,
       7.4658088e-02, 1.6295932e-02, 3.6165837e-02, 9.4850713e-01,
       4.2175108e-01, 1.7412608e-03, 6.7943684e-10, 9.6499955e-04,
       6.7725979e-02, 0.0000000e+00], dtype=float32), 'threshold': 0.0035714285714285718}, 'meta': {'name': 'TabularDrift', 'online': False, 'data_type': None, 'version': '0.11.4', 'detector_type': 'drift'}}


In [7]:
for idx in range(df.shape[1]):
  if preds['data']['p_val'][idx] < 0.05:
    print(df.columns[idx], preds['data']['p_val'][idx])

month 0.0
year 0.0
town 0.036623273
dist_to_nearest_stn 0.016295932
dist_to_dhoby 0.036165837
flat_model_type 0.0017412608
remaining_lease_years 6.7943684e-10
floor_area_sqm 0.00096499955
resale_price 0.0


> Assuming that the flurry of housing measures have made an impact on the relationship between all the features and resale_price (i.e. P(Y|X) changes), which type of data distribution shift possibly led to model degradation?


\# TODO: \
This is an example of covariate shift as the P(X) changes, but P(Y|X) remains the same. Covariate shifts can happen because of imbalanced datasets. This is further supported by the observation that data from the year 2023 is the least in the dataset whereas the majority of the data comes from year 2019.

> From your analysis via TabularDrift, which features contribute to this shift?


\# TODO: \
The categorical columns: 'flat_model_type', 'month' and 'year' have a p-value lower than the threshold of 0.5 and thus they cause this shift.

> Suggest 1 way to address model degradation and implement it, showing improved test R2 for year 2023.


\# TODO: \
One way to improve the the R^2 value is by training the model on a different set of data. In the new training, I trained the model on data from all the years except 2023. this significantly improved R^2 value from 0.16 to 0.55. We can also drop the columns of 'full_address' and 'nearest_stn' as more relevant metrics which are derived from them can be used to make the same conclusions.

In [8]:
# TODO: Enter your code here
from sklearn.preprocessing import StandardScaler, MinMaxScaler
from pytorch_tabular import TabularModel
from pytorch_tabular.models import CategoryEmbeddingModelConfig
from pytorch_tabular.config import (
    DataConfig,
    OptimizerConfig,
    TrainerConfig,
)

df = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/hdb_price_prediction.csv')

train_data = df[df['year'] <= 2022]
test_data = df[df['year'] == 2023]

train_data = train_data.drop(['full_address', 'nearest_stn', 'year'], axis=1)
test_data = test_data.drop(['full_address', 'nearest_stn', 'year'], axis=1)
# TODO: Enter your code here
data_config = DataConfig(
    target=["resale_price"],
    continuous_cols=["dist_to_nearest_stn", "dist_to_dhoby", "degree_centrality", "eigenvector_centrality", "remaining_lease_years", "floor_area_sqm"],
    categorical_cols=["month", "town", "flat_model_type", "storey_range"],
)


continuous_cols = ["dist_to_nearest_stn", "dist_to_dhoby", "degree_centrality", "eigenvector_centrality", "remaining_lease_years", "floor_area_sqm"]
scaler = StandardScaler()
train_data[continuous_cols] = scaler.fit_transform(train_data[continuous_cols])
test_data[continuous_cols] = scaler.transform(test_data[continuous_cols])


trainer_config = TrainerConfig(
    batch_size=1024,
    max_epochs=50,
    auto_lr_find=True
)

model_config = CategoryEmbeddingModelConfig(
    task="regression",
    layers="50",  # 1 hidden layer with 50 neurons
)

optimizer_config = OptimizerConfig(optimizer="Adam")

tabular_model = TabularModel(
    data_config=data_config,
    model_config=model_config,
    optimizer_config=optimizer_config,
    trainer_config=trainer_config
)

tabular_model.fit(train=train_data)
result = tabular_model.evaluate(test_data)[0]
pred_df = tabular_model.predict(test_data)

# Extract test loss and test MSE from the dictionary
test_loss = result['test_loss']
test_mse = result['test_mean_squared_error']

# Calculate RMSE
test_rmse = np.sqrt(test_mse)
print(f"Test RMSE: {test_rmse}")

# Calculate R-squared
test_target_values = test_data["resale_price"]
pred_values = pred_df["resale_price_prediction"]
test_r2 = 1 - (test_mse / np.var(test_target_values))
print(f"Test R^2: {test_r2}")

2023-10-14 15:14:50,047 - {pytorch_tabular.tabular_model:105} - INFO - Experiment Tracking is turned off
INFO:pytorch_tabular.tabular_model:Experiment Tracking is turned off
INFO:lightning_fabric.utilities.seed:Global seed set to 42
2023-10-14 15:14:50,117 - {pytorch_tabular.tabular_model:473} - INFO - Preparing the DataLoaders
INFO:pytorch_tabular.tabular_model:Preparing the DataLoaders
2023-10-14 15:14:50,147 - {pytorch_tabular.tabular_datamodule:290} - INFO - Setting up the datamodule for regression task
INFO:pytorch_tabular.tabular_datamodule:Setting up the datamodule for regression task
2023-10-14 15:14:51,711 - {pytorch_tabular.tabular_model:521} - INFO - Preparing the Model: CategoryEmbeddingModel
INFO:pytorch_tabular.tabular_model:Preparing the Model: CategoryEmbeddingModel
2023-10-14 15:14:51,841 - {pytorch_tabular.tabular_model:268} - INFO - Preparing the Trainer
INFO:pytorch_tabular.tabular_model:Preparing the Trainer
/usr/local/lib/python3.10/dist-packages/pytorch_lightning

Finding best initial lr:   0%|          | 0/100 [00:00<?, ?it/s]

/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called `self.log('train_loss', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called `self.log('train_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by doing `Trainer(logger=ALogger(...))`
  rank_zero_warn(
INFO:pytorch_lightning.utilities.rank_zero:`Trainer.fit` stopped: `max_steps=100` reached.
INFO:pytorch_lightning.tuner.lr_finder:Learning rate set to 0.5754399373371567
INFO:pytorch_lightning.utilities.rank_zero:Restoring states from the checkpoint path at /content/.lr_find_19afb099-b080-4967-bb06-31fa7af93066.ckpt
INFO:pytorch_lightning.utilities.rank_zero:Restored all states from the checkpoint file at /content/.lr_find_19afb099-b080-4967-bb06-31fa7af93066.ckpt
2023-10-14 

┏━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━┓
┃   ┃ Name             ┃ Type                      ┃ Params ┃
┡━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━┩
│ 0 │ _backbone        │ CategoryEmbeddingBackbone │  3.0 K │
│ 1 │ _embedding_layer │ Embedding1dLayer          │  1.6 K │
│ 2 │ head             │ LinearHead                │     51 │
│ 3 │ loss             │ MSELoss                   │      0 │
└───┴──────────────────┴───────────────────────────┴────────┘

Trainable params: 4.6 K                                                                                            
Non-trainable params: 0                                                                                            
Total params: 4.6 K                                                                                                
Total estimated model params size (MB): 0

Output()

2023-10-14 15:18:10,484 - {pytorch_tabular.tabular_model:584} - INFO - Training the model completed
INFO:pytorch_tabular.tabular_model:Training the model completed
2023-10-14 15:18:10,489 - {pytorch_tabular.tabular_model:1258} - INFO - Loading the best model
INFO:pytorch_tabular.tabular_model:Loading the best model


Output()

/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called 
`self.log('test_loss', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

/usr/local/lib/python3.10/dist-packages/pytorch_lightning/core/module.py:493: UserWarning: You called 
`self.log('test_mean_squared_error', ..., logger=True)` but have no logger configured. You can enable one by doing 
`Trainer(logger=ALogger(...))`
  rank_zero_warn(

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test_loss         │       13179405312.0       │
│  test_mean_squared_error  │       13179405312.0       │
└───────────────────────────┴───────────────────────────┘

Output()

Test RMSE: 114801.59106911367
Test R^2: 0.5529525422691766


### Appendix A



Here are our results from a linear regression model. We used StandardScaler for continuous variables and OneHotEncoder for categorical variables.

While 2021 data can be predicted well, test R2 dropped rapidly for 2022 and 2023.

| Training set | Test set | Test R2 |
|--------------|----------|---------|
| Year <= 2020 | 2021     | 0.76    |
| Year <= 2020 | **2022**     | 0.41    |
| Year <= 2020 | **2023**     | **0.10**   |



Similarly, a model trained on 2017 data can predict 2018-2021 well (with slight degradation in performance for 2021), but drops drastically in 2022 and 2023.

| Training set | Test set | Test R2 |
|--------------|----------|---------|
| 2017         | 2018     | 0.90    |
|              | 2019     | 0.89    |
|              | 2020     | 0.87    |
|              | 2021     | 0.72    |
|              | **2022**     | **0.37**    |
|              | **2023**     | **0.09**    |

With the test set fixed at year 2021, training on data from 2017-2020 still works well on the test data, with minimal degradation. Training sets closer to year 2021 generally do better.

| Training set | Test set | Test R2 |
|--------------|----------|---------|
| 2020         | 2021     | 0.81    |
| 2019         | 2021     | 0.75    |
| 2018         | 2021     | 0.73    |
| 2017         | 2021     | 0.72    |